internal/access/destinationcredential.go

```diff
@@ -0,0 +1,195 @@
+		ByDestination:              destination,
+		BySubject:                  uid.NewIdentityPolymorphicID(rCtx.Authenticated.User.ID),
+		IncludeInheritedFromGroups: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ListGrants: %w", err)
+	}
+	if len(grants) == 0 {
+		return nil, internal.ErrUnauthorized
+	}
+
+	dest, err := data.GetDestination(rCtx.DBTxn, data.GetDestinationOptions{ByName: destination})
+	if err != nil {
+		return nil, fmt.Errorf("GetDestination: %w", err)
+	}
+
+	cr := &models.DestinationCredential{
+		ID:                 uid.New(),
+		OrganizationMember: models.OrganizationMember{OrganizationID: rCtx.Authenticated.User.OrganizationID},
+		RequestExpiresAt:   time.Now().Add(2 * time.Minute),
+		UserID:             rCtx.Authenticated.User.ID,
+		DestinationID:      dest.ID,
+	}
+	err = data.CreateDestinationCredential(rCtx.DBTxn, cr)
+	if err != nil {
+		return nil, fmt.Errorf("CreateDestinationCredential: %w", err)
+	}
+	logging.Debugf("Creating DestinationCredential with id %d, orgID %d", cr.ID, cr.OrganizationID)
+
+	// commit the transaction so others can see the request
+	err = rCtx.DBTxn.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for the request to be filled; listen for the notify on the specific record req.ID
+	listener, err := data.ListenForNotify(c, rCtx.DataDB, data.ListenForNotifyOptions{
+		OrgID:                      rCtx.Authenticated.AccessKey.OrganizationID,
+		DestinationCredentialsByID: cr.ID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ListenForNotify: %w", err)
+	}
+	defer func() {
+		err := listener.Release(c)
+		if err != nil {
+			logging.Errorf("error releasing listener: %s", err)
+		}
+	}()
+
+	err = listener.WaitForNotification(c)
+	if err != nil {
+		logging.Debugf("CreateDestinationCredential notification timeout")
+		return nil, fmt.Errorf("WaitForNotification %w: %s", api.ErrTimeout, err)
+	}
+	logging.Debugf("CreateDestinationCredential notification received")
+
+	tx, err := rCtx.DataDB.Begin(rCtx.Request.Context(), &sql.TxOptions{
+		Isolation: sql.LevelReadCommitted,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("n: %w", err)
+	}
+	tx = tx.WithOrgID(cr.OrganizationID)
+
+	// looks like the DestinationCredential has been updated; reload the destination credential
+	logging.Debugf("Reloading DestinationCredential with id %d, orgID %d", cr.ID, cr.OrganizationID)
+	cr, err = data.GetDestinationCredential(tx, cr.ID, cr.OrganizationID)
+	if err != nil {
+		return nil, fmt.Errorf("GetDestinationCredential: %w", err)
+	}
+	if err = tx.Rollback(); err != nil {
+		logging.Debugf("could not roll back transaction")
+	}
+
+	return cr, nil
+}
+
+func ListDestinationCredentials(c *gin.Context, destination string, lastUpdateIndex int64) (ListDestinationCredentialResponse, error) {
+	rCtx := GetRequestContext(c)
+
+	roles := []string{models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole}
```

Comment:
This looks like something you should only be able to do for your own credentials, it looks like this could return bearer tokens

Reply:
it's actually only something the connector is supposed to be able to do. I'd love to tie this down to permissions that only the connector has.

Reply:
FYI the List function cannot be used to retrieve bearer tokens.